examples/voting/ballot.v.py

```diff
@@ -32,61 +41,88 @@ def __init__(_proposalNames: bytes32[2]):
+    # Throw if there is nothing to do:
+    assert self.voters[delegate_with_weight_to_forward].weight > 0
+
+    target = self.voters[delegate_with_weight_to_forward].delegate
+    for i in range(4):
+        if self.delegated(target):
+            target = self.voters[target].delegate
+            # The following effectively detects cycles of length <= 5.
+            # BUT cycles aren't actually problematic for correctness;
+            # they just result in spoiled votes.
+            # So, in the production version, this should instead be
+            # the responsibility of the contract's client, and this
+            # check should be removed.
+            assert target != delegate_with_weight_to_forward
+        else:
+            # weight will be moved to someone who directly voted or
+            # hasn't voted
+            break
+
+    weight_to_forward = self.voters[delegate_with_weight_to_forward].weight
+    self.voters[delegate_with_weight_to_forward].weight = 0
+    self.voters[target].weight += weight_to_forward
+
+    if self.directly_voted(target):
+        self.proposals[self.voters[target].vote].vote_count += weight_to_forward
+        self.voters[target].weight = 0
+
+    # To reiterate: if target is also a delegate, this function will need
+    # to be called again.
+
 # Delegate your vote to the voter `to`.
-def delegate(_to: address):
-    to = _to
+def delegate(to: address):
     # Throws if sender has already voted
-    assert not self.voters[msg.sender].voted 
-    # Throws if sender tries to delegate their vote to themselves
-    assert not msg.sender == to
-    # loop can delegate votes up to the current voter count
-    for i in range(self.voter_count, self.voter_count+1):
-        if self.voters[to].delegate:
-        # Because there are not while loops, use recursion to forward the delegation
-        # self.delegate(self.voters[to].delegate)
-            assert self.voters[to].delegate != msg.sender 
-            to = self.voters[to].delegate
+    assert not self.voters[msg.sender].voted
+    # Throws if sender tries to delegate their vote to themselves or to
+    # the default address value of 0x0000000000000000000000000000000000000000
+    # (the latter might not be problematic, but I don't want to think about it)
+    assert to != msg.sender and to != 0x0000000000000000000000000000000000000000
+
     self.voters[msg.sender].voted = True
     self.voters[msg.sender].delegate = to
-    if self.voters[to].voted:
-        # If the delegate already voted,
-        # directly add to the number of votes
-        self.proposals[self.voters[to].vote].vote_count += self.voters[msg.sender].weight
-    else:
-        # If the delegate did not vote yet,
-        # add to her weight.
-        self.voters[to].weight += self.voters[msg.sender].weight
+
+    # this call will throw iff this delegation would cause a loop of length <= 5.
+    self.forward_weight(msg.sender)
```

Comment:
`iff` should be `if`

Reply:
I meant `iff`. The two asserts at the beginning of `forward_weight` always evaluate to true when `forward_weight` is called by `delegate`. I mentioned this as a potential optimization enabled by the `should` keyword [here](https://github.com/ethereum/viper/issues/397#issuecomment-336508695).

Reply:
mathiness ftw! @DustinWehr maybe just spell out "if and only if"